021-bucket-folders.md

```diff
@@ -0,0 +1,223 @@
+
+## Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+**What problem are you trying to solve?**
+
+Large buckets become unmanageable. Users with hundreds or thousands of files have no way to organize them. Every file appears in a flat list. Finding specific files requires searching by name or using complex filtering.
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+### Design decisions
+
+1. **Folders are organizational** - No independent permissions, compression, or encryption settings
+2. **Flat structure initially** - No nested folders (folders can't contain folders)
+3. **Backward compatible** - Existing files stay at root level
+4. **Minimal overhead** - Folders are lightweight documents, not filesystem constructs
+
+### API Endpoints
+
+**POST /v1/storage/buckets/:bucketId/folders**
+Create a folder.
+```
+bucketId: string (required)
+folderId: string (required) - Custom ID or ID.unique()
+name: string (required) - Folder name (max 255 chars)
+permissions: array (optional) - Inherits bucket permissions if null
+```
+
+**GET /v1/storage/buckets/:bucketId/folders**
+List folders in a bucket.
+```
+bucketId: string (required)
+queries: array (optional) - Standard query support
+search: string (optional)
+```
+
+**GET /v1/storage/buckets/:bucketId/folders/:folderId**
+Get folder by ID.
+
+**PUT /v1/storage/buckets/:bucketId/folders/:folderId**
+Update folder name/permissions.
+
+**DELETE /v1/storage/buckets/:bucketId/folders/:folderId**
+Delete folder. Fails if folder contains files unless `force=true`.
+
+**Modified: POST /v1/storage/buckets/:bucketId/files**
+Add `folderId` parameter (optional, defaults to null = root level).
+
+**Modified: GET /v1/storage/buckets/:bucketId/files**
+Add `folderId` query parameter to filter files by folder.
+
+### Data Structure
+
+**New: Folders collection**
+Stored in same `bucket_{sequence}` collection as files, differentiated by type.
+
+```php
+// Folder document
+[
+    '$id' => 'unique_folder_id',
+    'type' => 'folder', // discriminator field
+    'bucketId' => 'bucket_id',
+    'bucketInternalId' => 123,
+    'name' => 'Invoices',
+    '$permissions' => [...], // same as bucket unless overridden
+    'search' => 'folder_id Invoices',
+    'filesCount' => 0, // cached count
+]
+
+// Modified file document - add single field
+[
+    // ... all existing fields ...
+    'folderId' => 'folder_id_or_null', // null = root level
+]
+```
+
+**Required indexes:**
+- `(bucketInternalId, type, name)` - unique constraint for folder names
+- `(bucketInternalId, folderId)` - list files in folder
+- `(bucketInternalId, type)` - list all folders
+
+**Migration:**
+1. Add `folderId` field to all existing file documents (default null)
```

Comment:
We should get DB team (specifically Shmuel to help with this), we have a few migrations upcoming, could save time.